src/server/tasks/export-multiple-campaigns.ts

```diff
@@ -0,0 +1,191 @@
+    .reader("campaign_contact")
+    .count("*")
+    .where({ campaign_id: campaignId });
+  const contactsCount = typeof count === "number" ? count : parseInt(count, 10);
+
+  // Attempt upload to cloud storage
+  const fileNameKey = campaignTitle.replace(/ /g, "_").replace(/\//g, "_");
+  const timestamp = DateTime.local().toFormat("y-mm-d-hh-mm-ss");
+  const keyWithTimeStamp = `${fileNameKey}-${timestamp}`;
+
+  const campaignMetaData = {
+    fileNameKey: keyWithTimeStamp,
+    campaignId,
+    campaignTitle,
+    contactsCount,
+    helpers,
+    interactionSteps,
+    campaignVariableNames
+  };
+
+  const exportUrls = await processExportData(campaignMetaData, spokeOptions);
+  helpers.logger.info(
+    `Exported data for campaign: ${campaignTitle} - ID ${campaignId}`
+  );
+
+  // store exportUrls in job_request table
+  await helpers.updateResult({ message: JSON.stringify(exportUrls) });
+  // indicate that the exports have been processed
+  await helpers.updateStatus(100);
+};
+
+type EmailBulkOperationPayload = ExportCampaignPayload & {
+  jobRequestRecords: JobRequestRecord[];
+  campaignIds: string[];
+};
+
+// eslint-disable-next-line max-len
+export const sendEmailForBulkExportOperation: ProgressTask<EmailBulkOperationPayload> = async (
+  payload,
+  helpers
+) => {
+  const { jobRequestRecords, campaignId, requesterId, campaignIds } = payload;
+
+  // map campaign id to campaign title and export urls for email composition
+  const campaignMetaDataMap: CampaignExportMetaData = {};
+  for (const id of campaignIds) {
+    const parsed = parseInt(id, 10);
+    const { campaignTitle } = await fetchExportData(parsed, requesterId);
+    campaignMetaDataMap[id] = { campaignTitle, exportUrls: null };
+  }
+
+  const jobRequestIds = jobRequestRecords.map((record) => record.id);
+  // query job_req table for campaign export urls where status is 100 (exports complete)
+  const { rows } = await helpers.query(
+    `
+      select campaign_id, result_message 
+      from job_request 
+      where id = ANY ($1)
+        and status = 100
+    `,
+    [jobRequestIds]
+  );
+
+  // wait for all campaign exports to process
+  const exportsStillProcessing = rows.length !== campaignIds.length;
+
+  // map query result to campaign id and parse JSON
+  const campaignExportsMap = rows.map((result) => ({
+    campaignId: result.campaign_id,
+    exportUrls: JSON.parse(result.result_message)
+  }));
+
+  // map fetched export urls to  campaignMetaData
+  for (const campaignExport of campaignExportsMap) {
+    if (
+      !Object.keys(campaignMetaDataMap).includes(
+        campaignExport.campaignId.toString()
+      )
+    ) {
+      throw new Error("attempted to store exportUrls for incorrect campaign");
+    }
+    campaignMetaDataMap[campaignExport.campaignId].exportUrls =
+      campaignExport.exportUrls;
+  }
+
+  try {
+    if (exportsStillProcessing) {
+      throw new Error("Attempting to send export email before expots process");
+    }
```

Comment:
@bchrobot Any thoughts on how this job might execute polling better? I thought that throwing if the `rows` returned from the query were not the same length as the `campaignIds` included in the job payload would be enough to re-try the job until it succeeds, but so far in testing it's succeeding indiscriminately. It's also hard to predict when the job will re-run, even with a `priority: 0` in the `task-spec`

Reply:
Ohhh, nice - thanks! I'll try that approach